test/python/dagcircuit/test_dagcircuit.py

```diff
@@ -1226,6 +1226,62 @@ def test_circuit_factors(self):
         """Test number of separable factors in circuit."""
         self.assertEqual(self.dag.num_tensor_factors(), 2)
 
+    def test_separable_circuits(self):
+        """Test separating disconnected sets of qubits in a circuit."""
+        # Empty case
+        dag = DAGCircuit()
+        self.assertEqual([], dag.separable_circuits())
+
+        # 3 disconnected qubits
+        qreg = QuantumRegister(3, "qr")
+        creg = ClassicalRegister(2, "cr")
+        dag.add_qreg(qreg)
+        dag.add_creg(creg)
+        dag.apply_operation_back(HGate(), [qreg[0]], [])
+        dag.apply_operation_back(HGate(), [qreg[1]], [])
+        dag.apply_operation_back(HGate(), [qreg[2]], [])
+        dag.apply_operation_back(HGate(), [qreg[2]], [])
+        self.assertEqual(3, len(dag.separable_circuits(remove_idle_wires=True)))
+
+        # 2 sets of disconnected qubits
+        dag.apply_operation_back(CXGate(), [qreg[1], qreg[2]], [])
+        self.assertEqual(2, len(dag.separable_circuits(remove_idle_wires=True)))
+
+        # One connected component
+        dag.apply_operation_back(CXGate(), [qreg[0], qreg[1]], [])
+        self.assertEqual(1, len(dag.separable_circuits(remove_idle_wires=True)))
+
+        # Test circuit ordering with measurements
+        dag = DAGCircuit()
+        qreg = QuantumRegister(3, "q")
+        creg = ClassicalRegister(1, "c")
+        dag.add_qreg(qreg)
+        dag.add_creg(creg)
+        dag.apply_operation_back(HGate(), [qreg[0]], [])
+        dag.apply_operation_back(XGate(), [qreg[0]], [])
+        dag.apply_operation_back(XGate(), [qreg[1]], [])
+        dag.apply_operation_back(HGate(), [qreg[1]], [])
+        dag.apply_operation_back(YGate(), [qreg[2]], [])
+        dag.apply_operation_back(HGate(), [qreg[2]], [])
+        dag.apply_operation_back(Measure(), [qreg[0]], [creg[0]])
+
+        compare1 = QuantumCircuit(3, 1)
+        compare1.h(0)
+        compare1.x(0)
+        compare1.measure(0, 0)
+        compare2 = QuantumCircuit(3, 1)
+        compare2.x(1)
+        compare2.h(1)
+        compare3 = QuantumCircuit(3, 1)
+        compare3.y(2)
+        compare3.h(2)
+
+        dags = dag.separable_circuits()
+
+        self.assertEqual(dag_to_circuit(dags[0]), compare1)
+        self.assertEqual(dag_to_circuit(dags[1]), compare2)
+        self.assertEqual(dag_to_circuit(dags[2]), compare3)
```

Comment:
I think this one introduces a completely disjoint change, right?  It would be good to put that in a separate test case, so they both run, even if one fails.

Also, it'd be better to construct the `compare1`, `compare2` and `compare3` elements as a list of `DAGCircuit`, and then compare the two lists directly - that way we get more information on how many / which circuits fail, if there's a failure.  I'm also a bit nervous that this test _requires_ that the output has a particular order, but I don't think `rustworkx.weakly_connect_components` will guarantee that for us?  It might be ok to leave for now, but if there's an easy way to make the test resistant to re-orderings of the list, that would be good.

Reply:
> I don't think `rustworkx.weakly_connect_components` will guarantee that for us?

I asked about `rustworkx.connected_components` [on Qiskit Slack](https://qiskit.slack.com/archives/C0241HXNWNM/p1681157089998819), and here is what Matthew said:

> There isn't an explicit guarantee to the order, mostly because I don't think we want to commit to it being stable in case we want to change the internal implementation of the algorithm in the future.

In that case, it was easy to make the operation deterministic by calling `sort` or the result.  Something similar may be possible here.

Reply:
That's a good idea, thanks Jim.  I think we could find a way to choose the "first" busy qubit in each returned DAG, and sort the final list based on that qubit's index into the parent DAG.  Given the nature of the function, that should be a unique sort key, so should deterministically order the list.

Just to be clear: I think we only need to bother with that in the test; I don't think we need to commit to any particular order in the method itself, just like `weakly_connected_components` doesn't.